File: packages/sequent-core/src/types/results.rs
// SPDX-FileCopyrightText: 2025 Sequent Tech Inc <legal@sequentech.io>
//
// SPDX-License-Identifier: AGPL-3.0-only
#![allow(non_camel_case_types)]
use chrono::{DateTime, Local};
use ordered_float::NotNan;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::default::Default;

// Keys for annotations fields in ResultAreaContest
pub const EXTENDED_METRICS: &str = "extended_metrics";
pub const PROCESS_RESULTS: &str = "process_results";

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub enum ResultDocumentType {
    Json,
    Pdf,
    Html,
    TarGz,
    TarGzOriginal,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, PartialEq, Eq)]
pub struct ResultDocuments {
    pub json: Option<String>,
    pub pdf: Option<String>,
    pub html: Option<String>,
    pub tar_gz: Option<String>,
    pub tar_gz_original: Option<String>,
    pub tar_gz_pdfs: Option<String>,
    pub all_areas_html: Option<String>,
    pub all_areas_json: Option<String>,
}

impl ResultDocuments {
    pub fn get_document_by_type(
        &self,
        doc_type: &ResultDocumentType,
    ) -> Option<String> {
        match doc_type {
            ResultDocumentType::Json => self.json.clone(),
            ResultDocumentType::Pdf => self.pdf.clone(),
            ResultDocumentType::Html => self.html.clone(),
            ResultDocumentType::TarGz => self.tar_gz.clone(),
            ResultDocumentType::TarGzOriginal => self.tar_gz_original.clone(),
        }
    }
}

#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
pub struct ResultsEvent {
    pub id: String,
    pub tenant_id: String,
    pub election_event_id: String,
    pub name: Option<String>,
    pub created_at: Option<DateTime<Local>>,
    pub last_updated_at: Option<DateTime<Local>>,
    pub labels: Option<Value>,
    pub annotations: Option<Value>,
    pub documents: Option<ResultDocuments>,
}

#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
pub struct ResultsElection {
    pub id: String,
    pub tenant_id: String,
    pub election_event_id: String,
    pub election_id: String,
    pub results_event_id: String,
    pub name: Option<String>,
    pub elegible_census: Option<i64>,
    pub total_voters: Option<i64>,
    pub created_at: Option<DateTime<Local>>,
    pub last_updated_at: Option<DateTime<Local>>,
    pub labels: Option<Value>,
    pub annotations: Option<Value>,
    pub total_voters_percent: Option<NotNan<f64>>,
    pub documents: Option<ResultDocuments>,
}

#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
pub struct ResultsElectionArea {
    pub id: String,
    pub tenant_id: String,
    pub election_event_id: String,
    pub election_id: String,
    pub area_id: String,
    pub results_event_id: String,
    pub created_at: Option<DateTime<Local>>,
    pub last_updated_at: Option<DateTime<Local>>,
    pub documents: Option<ResultDocuments>,
    pub name: Option<String>,
}

#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
pub struct ResultsContest {
    pub id: String,
    pub tenant_id: String,
    pub election_event_id: String,
    pub election_id: String,
    pub contest_id: String,
    pub results_event_id: String,
    pub elegible_census: Option<i64>,
    /// Ballots that are not invalid and not declined. Explicit and implicit
    /// blank ballots are included; selecting explicit blank with a regular
    /// candidate is an implicit invalid ballot.
    pub total_valid_votes: Option<i64>,
    pub explicit_invalid_votes: Option<i64>,
    pub implicit_invalid_votes: Option<i64>,
    #[serde(alias = "blank_votes")]
    pub total_blank_votes: Option<i64>,
    pub explicit_blank_votes: Option<i64>,
    pub implicit_blank_votes: Option<i64>,
    pub voting_type: Option<String>,
    pub counting_algorithm: Option<String>,
    pub name: Option<String>,
    pub created_at: Option<DateTime<Local>>,
    pub last_updated_at: Option<DateTime<Local>>,
    pub labels: Option<Value>,
    pub annotations: Option<Value>,
    pub total_invalid_votes: Option<i64>,
    pub total_invalid_votes_percent: Option<NotNan<f64>>,
    pub total_valid_votes_percent: Option<NotNan<f64>>,
    pub explicit_invalid_votes_percent: Option<NotNan<f64>>,
    pub implicit_invalid_votes_percent: Option<NotNan<f64>>,
    #[serde(alias = "blank_votes_percent")]
    pub total_blank_votes_percent: Option<NotNan<f64>>,
    pub explicit_blank_votes_percent: Option<NotNan<f64>>,
    pub implicit_blank_votes_percent: Option<NotNan<f64>>,
    pub total_votes: Option<i64>,
    pub total_votes_percent: Option<NotNan<f64>>,
    pub documents: Option<ResultDocuments>,
    pub total_auditable_votes: Option<i64>,
    pub total_auditable_votes_percent: Option<NotNan<f64>>,
}

#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
pub struct ResultsContestCandidate {
    pub id: String,
    pub tenant_id: String,
    pub election_event_id: String,
    pub election_id: String,
    pub contest_id: String,
    pub candidate_id: String,
    pub results_event_id: String,
    pub cast_votes: Option<i64>,
    pub winning_position: Option<i64>,
    pub points: Option<i64>,
    pub created_at: Option<DateTime<Local>>,
    pub last_updated_at: Option<DateTime<Local>>,
    pub labels: Option<Value>,
    pub annotations: Option<Value>,
    pub cast_votes_percent: Option<NotNan<f64>>,
    pub documents: Option<ResultDocuments>,
}

#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
pub struct ResultsAreaContest {
    pub id: String,
    pub tenant_id: String,
    pub election_event_id: String,
    pub election_id: String,
    pub contest_id: String,
    pub area_id: String,
    pub results_event_id: String,
    pub elegible_census: Option<i64>,
    /// Ballots that are not invalid and not declined. Explicit and implicit
    /// blank ballots are included; selecting explicit blank with a regular
    /// candidate is an implicit invalid ballot.
    pub total_valid_votes: Option<i64>,
    pub explicit_invalid_votes: Option<i64>,
    pub implicit_invalid_votes: Option<i64>,
    #[serde(alias = "blank_votes")]
    pub total_blank_votes: Option<i64>,
    pub explicit_blank_votes: Option<i64>,
    pub implicit_blank_votes: Option<i64>,
    pub created_at: Option<DateTime<Local>>,
    pub last_updated_at: Option<DateTime<Local>>,
    pub labels: Option<Value>,
    pub annotations: Option<Value>,
    pub total_valid_votes_percent: Option<NotNan<f64>>,
    pub total_invalid_votes: Option<i64>,
    pub total_invalid_votes_percent: Option<NotNan<f64>>,
    pub explicit_invalid_votes_percent: Option<NotNan<f64>>,
    pub implicit_invalid_votes_percent: Option<NotNan<f64>>,
    #[serde(alias = "blank_votes_percent")]
    pub total_blank_votes_percent: Option<NotNan<f64>>,
    pub explicit_blank_votes_percent: Option<NotNan<f64>>,
    pub implicit_blank_votes_percent: Option<NotNan<f64>>,
    pub total_votes: Option<i64>,
    pub total_votes_percent: Option<NotNan<f64>>,
    pub documents: Option<ResultDocuments>,
    pub total_auditable_votes: Option<i64>,
    pub total_auditable_votes_percent: Option<NotNan<f64>>,
}

#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
pub struct ResultsAreaContestCandidate {
    pub id: String,
    pub tenant_id: String,
    pub election_event_id: String,
    pub election_id: String,
    pub contest_id: String,
    pub area_id: String,
    pub candidate_id: String,
    pub results_event_id: String,
    pub cast_votes: Option<i64>,
    pub winning_position: Option<i64>,
    pub points: Option<i64>,
    pub created_at: Option<DateTime<Local>>,
    pub last_updated_at: Option<DateTime<Local>>,
    pub labels: Option<Value>,
    pub annotations: Option<Value>,
    pub cast_votes_percent: Option<NotNan<f64>>,
    pub documents: Option<ResultDocuments>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn legacy_blank_vote_fields_deserialize_into_renamed_totals() {
        let contest: ResultsContest =
            serde_json::from_value(serde_json::json!({
                "id": "result",
                "tenant_id": "tenant",
                "election_event_id": "event",
                "election_id": "election",
                "contest_id": "contest",
                "results_event_id": "results",
                "blank_votes": 7,
                "blank_votes_percent": 0.25
            }))
            .unwrap();
        assert_eq!(contest.total_blank_votes, Some(7));
        assert_eq!(
            contest.total_blank_votes_percent.map(NotNan::into_inner),
            Some(0.25)
        );
        assert_eq!(contest.explicit_blank_votes, None);
        assert_eq!(contest.implicit_blank_votes, None);

        let area_contest: ResultsAreaContest =
            serde_json::from_value(serde_json::json!({
                "id": "result",
                "tenant_id": "tenant",
                "election_event_id": "event",
                "election_id": "election",
                "contest_id": "contest",
                "area_id": "area",
                "results_event_id": "results",
                "blank_votes": 3,
                "blank_votes_percent": 0.5
            }))
            .unwrap();
        assert_eq!(area_contest.total_blank_votes, Some(3));
        assert_eq!(
            area_contest
                .total_blank_votes_percent
                .map(NotNan::into_inner),
            Some(0.5)
        );
        assert_eq!(area_contest.explicit_blank_votes, None);
        assert_eq!(area_contest.implicit_blank_votes, None);
    }
}

File: packages/results-portal/src/types/results.ts
// SPDX-FileCopyrightText: 2026 Sequent Tech Inc <legal@sequentech.io>
//
// SPDX-License-Identifier: AGPL-3.0-only

export type ResultsAccess = "public" | "authenticated"
export type ResultsVisibilityScope = "full_event" | "area_based"
export type ResultsRouteScope = "event" | "election"
export type ContestPublicationState = "published" | "not_published"
export type ResultsSqlId = string | number
export type ResultsSqlNumber = string | number | null
export type ResultsJson =
    | string
    | number
    | boolean
    | null
    | ResultsJson[]
    | {[key: string]: ResultsJson}
export type ResultsSerializedJson = string | ResultsJson

export interface ResultsPublicationIndexEntry {
    publication_id: string
    route_scope: ResultsRouteScope
    route?: string | null
    route_election_id?: string | null
    election_ids?: string[]
    access: ResultsAccess
    visibility_scope?: ResultsVisibilityScope
    manifest_public_path?: string | null
    manifest_url?: string | null
}

export interface ResultsPublicationIndex {
    schema_version: number
    tenant_id?: string
    election_event_id?: string
    publications?: ResultsPublicationIndexEntry[]
}

export interface ResultsManifestContest {
    election_id: string
    contest_id: string
    area_id?: string | null
    publication_state: ContestPublicationState
    positions?: number | null
}

export interface ResultsManifestArtifact {
    public_path?: string | null
    url?: string | null
    document_id?: string | null
}

export interface ResultsManifestCustomCss {
    election_event?: string | null
    elections?: Record<string, string | null>
}

export interface ResultsManifest {
    schema_version: number
    tenant_id: string
    election_event_id: string
    election_ids: string[]
    route_scope: ResultsRouteScope
    route_election_id?: string | null
    publication_id: string
    tally_session_id?: string
    tally_session_execution_id?: string
    results_event_id: string
    version: number
    published_at?: string
    access: ResultsAccess
    visibility_scope: ResultsVisibilityScope
    default_locale?: string
    available_languages?: string[]
    title?: Record<string, string> | string
    custom_css?: ResultsManifestCustomCss
    contests: ResultsManifestContest[]
    artifacts: {
        full_sqlite?: ResultsManifestArtifact | null
        areas?: Record<string, ResultsManifestArtifact> | null
    }
}

export interface ResultsResolverResponse {
    tenant_id: string
    election_event_id: string
    access: ResultsAccess
    route_scope: ResultsRouteScope
    election_ids: string[]
    publication_id: string
    manifest_public_path?: string | null
    manifest_url?: string | null
    manifest?: ResultsManifest | null
}

export interface ResultsRow {
    id?: ResultsSqlId | null
    election_event_id?: ResultsSqlId | null
    election_id?: ResultsSqlId | null
    contest_id?: ResultsSqlId | null
    candidate_id?: ResultsSqlId | null
    area_id?: ResultsSqlId | null
    results_event_id?: ResultsSqlId | null
    name?: string | null
    description?: string | null
    type?: string | null
    labels?: ResultsSerializedJson
    presentation?: ResultsSerializedJson
    annotations?: ResultsSerializedJson
    external_id?: string | null
    is_public?: boolean | number | null
    is_acclaimed?: boolean | number | null
    is_active?: boolean | number | null
    min_votes?: ResultsSqlNumber
    max_votes?: ResultsSqlNumber
    winning_candidates_num?: ResultsSqlNumber
    voting_type?: string | null
    counting_algorithm?: string | null
    elegible_census?: ResultsSqlNumber
    total_voters?: ResultsSqlNumber
    total_voters_percent?: ResultsSqlNumber
    total_valid_votes?: ResultsSqlNumber
    total_valid_votes_percent?: ResultsSqlNumber
    total_invalid_votes?: ResultsSqlNumber
    total_invalid_votes_percent?: ResultsSqlNumber
    explicit_invalid_votes?: ResultsSqlNumber
    explicit_invalid_votes_percent?: ResultsSqlNumber
    implicit_invalid_votes?: ResultsSqlNumber
    implicit_invalid_votes_percent?: ResultsSqlNumber
    total_blank_votes?: ResultsSqlNumber
    total_blank_votes_percent?: ResultsSqlNumber
    blank_votes?: ResultsSqlNumber
    blank_votes_percent?: ResultsSqlNumber
    explicit_blank_votes?: ResultsSqlNumber
    explicit_blank_votes_percent?: ResultsSqlNumber
    implicit_blank_votes?: ResultsSqlNumber
    implicit_blank_votes_percent?: ResultsSqlNumber
    total_votes?: ResultsSqlNumber
    total_votes_percent?: ResultsSqlNumber
    total_auditable_votes?: ResultsSqlNumber
    total_auditable_votes_percent?: ResultsSqlNumber
    cast_votes?: ResultsSqlNumber
    cast_votes_percent?: ResultsSqlNumber
    winning_position?: ResultsSqlNumber
    points?: ResultsSqlNumber
}

export interface ResultsElectionRow extends ResultsRow {
    election_id: ResultsSqlId
}

export interface ResultsElectionAreaRow extends ResultsElectionRow {
    area_id: ResultsSqlId
}

export interface ResultsContestRow extends ResultsElectionRow {
    contest_id: ResultsSqlId
}

export interface ResultsAreaContestRow extends ResultsContestRow {
    area_id: ResultsSqlId
}

export interface ResultsContestCandidateRow extends ResultsContestRow {
    candidate_id: ResultsSqlId
}

export interface ResultsAreaContestCandidateRow extends ResultsContestCandidateRow {
    area_id: ResultsSqlId
}

export interface ResultsSqliteDataset {
    election_event: ResultsRow[]
    election: ResultsRow[]
    contest: ResultsRow[]
    candidate: ResultsRow[]
    area: ResultsRow[]
    results_event: ResultsRow[]
    results_election: ResultsElectionRow[]
    results_election_area: ResultsElectionAreaRow[]
    results_contest: ResultsContestRow[]
    results_contest_candidate: ResultsContestCandidateRow[]
    results_area_contest: ResultsAreaContestRow[]
    results_area_contest_candidate: ResultsAreaContestCandidateRow[]
}

const isObject = (value: unknown): value is {[key: string]: unknown} =>
    typeof value === "object" && value !== null && !Array.isArray(value)

const isStringArray = (value: unknown): value is string[] =>
    Array.isArray(value) && value.every((item) => typeof item === "string")

const isOptionalString = (value: unknown): value is string | null | undefined =>
    value === undefined || value === null || typeof value === "string"

const isResultsAccess = (value: unknown): value is ResultsAccess =>
    value === "public" || value === "authenticated"

const isResultsVisibilityScope = (value: unknown): value is ResultsVisibilityScope =>
    value === "full_event" || value === "area_based"

const isResultsRouteScope = (value: unknown): value is ResultsRouteScope =>
    value === "event" || value === "election"

const isPublicationIndexEntry = (value: unknown): value is ResultsPublicationIndexEntry =>
    isObject(value) &&
    typeof value.publication_id === "string" &&
    isResultsRouteScope(value.route_scope) &&
    isResultsAccess(value.access) &&
    isOptionalString(value.route) &&
    isOptionalString(value.route_election_id) &&
    (value.election_ids === undefined || isStringArray(value.election_ids)) &&
    (value.visibility_scope === undefined || isResultsVisibilityScope(value.visibility_scope)) &&
    isOptionalString(value.manifest_public_path) &&
    isOptionalString(value.manifest_url)

export const parseResultsPublicationIndex = (value: unknown): ResultsPublicationIndex => {
    if (isPublicationIndexEntry(value)) {
        return {schema_version: 1, publications: [value]}
    }

    if (
        !isObject(value) ||
        typeof value.schema_version !== "number" ||
        (value.publications !== undefined &&
            (!Array.isArray(value.publications) ||
                !value.publications.every(isPublicationIndexEntry)))
    ) {
        throw new Error("Invalid results publication index")
    }

    return value as ResultsPublicationIndex
}

const isManifestContest = (value: unknown): value is ResultsManifestContest =>
    isObject(value) &&
    typeof value.election_id === "string" &&
    typeof value.contest_id === "string" &&
    (value.publication_state === "published" || value.publication_state === "not_published")

const isManifestArtifact = (value: unknown): value is ResultsManifestArtifact =>
    isObject(value) &&
    isOptionalString(value.public_path) &&
    isOptionalString(value.url) &&
    isOptionalString(value.document_id)

const isManifestArtifactMap = (value: unknown): value is Record<string, ResultsManifestArtifact> =>
    isObject(value) && Object.values(value).every(isManifestArtifact)

const isManifestArtifacts = (value: unknown): value is ResultsManifest["artifacts"] =>
    isObject(value) &&
    (value.full_sqlite === undefined ||
        value.full_sqlite === null ||
        isManifestArtifact(value.full_sqlite)) &&
    (value.areas === undefined || value.areas === null || isManifestArtifactMap(value.areas))

export const parseResultsManifest = (value: unknown): ResultsManifest => {
    if (
        !isObject(value) ||
        typeof value.schema_version !== "number" ||
        typeof value.tenant_id !== "string" ||
        typeof value.election_event_id !== "string" ||
        !isStringArray(value.election_ids) ||
        !isResultsRouteScope(value.route_scope) ||
        typeof value.publication_id !== "string" ||
        typeof value.results_event_id !== "string" ||
        typeof value.version !== "number" ||
        !isResultsAccess(value.access) ||
        !isResultsVisibilityScope(value.visibility_scope) ||
        !Array.isArray(value.contests) ||
        !value.contests.every(isManifestContest) ||
        !isManifestArtifacts(value.artifacts)
    ) {
        throw new Error("Invalid results publication manifest")
    }

    return value as ResultsManifest
}
